content/advent-2019/pkgapi.md

```diff
@@ -0,0 +1,1922 @@
+
+The `willshakes` library is an example of a resource library. For a real life use case, I offer the [mnist](gorgonia.org/gorgonia/examples/mnist) library for consideration. Due to the nature of the source data, as well as the inability for Go to handle static resources, the design of the package is limited to source code that loads data from an external file into a data structure. In the Python world, this is a different case as one may use [`keras.datasets.mnist`](https://keras.io/datasets/) immediately as a resource.
+
+
+
+
+
+# What Makes A Good Library? #
+
+Now that I have defined the things that can go into a library, let's take a look at what makes a good library. We will start with the simpler, more obvious statements, before moving on to more nuanced considerations. Despite this, the simpler statements often come with caveats, which will be briefly explored.
+
+I have a few principles that form the basic principles of what makes a good library:
+
+1. Reliable
+2. Easy to use/build
+3. Generic
+
+## Reliable ##
+
+First and foremost, a library must be reliable. What is the point of using a library if it cannot reliably do what it claims to do? There are many features of a  reliable library, enumerated below.
+
+### Does One Thing ###
+
+A good library does one thing or provides one resource. What constitute "one thing" is usually the point of contention.
+
+For example, consider the [grpc](https://godoc.org/google.golang.org/grpc) library. It does one thing - gRPC. But gRPC has many subcomponents to it - server and client are the two primary subcomponents.
+
+An example on the other extreme can be seen in the packages that pervade npm. `left-pad` was a package that provided one function that padded a string. It did one thing, and many packages depended upon it. Thus when the `left-pad` package was unpublished, it broke the internet.
+
+
+### Is Well-Tested ###
+
+A good library is well-tested. Users who use the library must be able to feel confident about the library they're using. I would go so far to prefer only libraries that have been tested using property-based testing (I had previously written [an article about property-based testing on GopherAcademy](https://blog.gopheracademy.com/advent-2017/property-based-testing/)) or have been fuzz-tested.
+
+A good thing to check on a well-tested library is whether the tests test for general cases or only specific cases. This is why I prefer libraries that are fuzz-tested and have PBTs in them. Fuzz-testing checks that the libary functions can handle unforeseen input, while property-based testing requires a deep understanding of the domain space.
+
+Having said that, if you develop driver libaraies, it might be a bit difficult to test such libraries. Perhaps this is an indictment on my poor ability to reason about testing, but I have found no good general-purpose testing patterns in the case of driver libraries.
```

Comment:
You can see some ideas in https://www.youtube.com/watch?v=xzG0pLM4Q64. but in general - coverage guided fuzzing has been pretty successful for breaking driver libraries / drivers.